torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,401 @@
+using nss = std::chrono::nanoseconds;
+
+std::atomic<uint32_t> TensorPipeAgent::nextShmAddressID_(0);
+
+TensorPipeAgent::TensorPipeAgent(
+    worker_id_t selfId,
+    std::string selfName,
+    std::shared_ptr<::c10d::Store> addressStore,
+    TensorPipeRpcBackendOptions opts)
+    : RpcAgent(
+          WorkerInfo(std::move(selfName), std::move(selfId)),
+          std::make_unique<RequestCallbackImpl>(),
+          opts.rpcTimeout),
+      context_(std::make_shared<tensorpipe::Context>()),
+      addressStore_(std::move(addressStore)),
+      opts_(std::move(opts)) {
+  // Generate the maps for once.
+  for (const std::pair<std::string, worker_id_t>& kv : opts_.workerNameToId) {
+    const string& workerName = kv.first;
+    worker_id_t workerId = kv.second;
+    workerIdToInfo_.emplace(workerId, WorkerInfo(workerName, workerId));
+    workerNameToInfo_.emplace(workerName, WorkerInfo(workerName, workerId));
+  }
+}
+
+TensorPipeAgent::~TensorPipeAgent() {
+  shutdownImpl();
+}
+
+void TensorPipeAgent::startImpl() {
+  context_->registerTransport(
+      1, "tcp", std::make_shared<tensorpipe::transport::uv::Context>());
+#ifdef TP_ENABLE_SHM
+  context_->registerTransport(
+      0, "shm", std::make_shared<tensorpipe::transport::shm::Context>());
+#endif
+  context_->registerChannel(
+      1, "basic", std::make_shared<tensorpipe::channel::basic::Context>());
+#ifdef TP_ENABLE_CMA
+  context_->registerChannel(
+      0, "cma", std::make_shared<tensorpipe::channel::cma::Context>());
+#endif
+
+  std::vector<std::string> addresses = {createUniqueShmAddr(),
+                                        "tcp://127.0.0.1"};
+
+  listener_ = context_->listen(addresses);
+
+  // Store our own url.
+  const auto address = listener_->url("tcp");
+  const std::vector<uint8_t> selfAddrData(address.begin(), address.end());
+  addressStore_->set(workerInfo_.name_, selfAddrData);
+
+  for (const auto& p : workerNameToInfo_) {
+    const auto& name = p.first;
+    auto nodeAddrData = addressStore_->get(name);
+    auto nodeAddrStr =
+        std::string((const char*)nodeAddrData.data(), nodeAddrData.size());
+    workerNameToURL_.insert({name, nodeAddrStr});
+  }
+
+  listener_->accept([this](
+                        const tensorpipe::Error& error,
+                        std::shared_ptr<tensorpipe::Pipe> pipe) {
+    onListenerAccepted(error, pipe);
+  });
+
+  isRunning_ = true;
+}
+
+void TensorPipeAgent::onListenerAccepted(
+    const tensorpipe::Error& error,
+    std::shared_ptr<tensorpipe::Pipe>& pipe) {
+  // TODO: Handle listener error
+  if (error) {
+    LOG(WARNING) << "got error from listener: " << error.what();
+    return;
+  }
+
+  // Accept the next connection request
+  listener_->accept([this](
+                        const tensorpipe::Error& error,
+                        std::shared_ptr<tensorpipe::Pipe> pipe) {
+    onListenerAccepted(error, pipe);
+  });
+
+  // Arm for server read
+  respond(pipe);
```

Comment:
We can see TensorPipe as single loop thread. Think of a socket. It's designed to create multiple pipes (pipe pool) to achieve better performance. In current solution, say we have trainer:0 sending rpc requests to es:0:0-9, there would be 10 pipes. To bump a single pair of trainer:0 to es:0:0, we can have a pool of pipes. I'll add a TODO for this

Reply:
Another option is to have the single loop thread only accept connections and write responses back to the socket. For deserialization of request, calling RequestCallbackImpl and serialization of response, is it possible to defer this work to a separate threadpool?

Reply:
> Another option is to have the single loop thread only accept connections and write responses back to the socket. For deserialization of request, calling RequestCallbackImpl and serialization of response, is it possible to defer this work to a separate threadpool?

I suspect that such an approach would be trickier to use here, since the tensorpipes approach attempts to read directly into the result tensor buffer - serialization is more tied with IO than e.g. in the Thrift approach (but OTOH tensorpipe deserialization is somewhat lighter weight)

If we need more bandwidth between a pair of endpoints than a single thread allows, ideally we might have a parallel option to Thrift's PooledRequestChannel here, where a connection pair can effectively "trunk" across a set of threads/sockets (and avoid some head-of-line-blocking issues, though giving up fifo/ordering)